api/users.go

```diff
@@ -53,35 +60,383 @@ func (self *ApiServer) SetPassword(
+		return nil, err
+	} else if user_record != nil && createUser {
+		return nil, status.Errorf(codes.AlreadyExists,
+					  "Cannot create user %s.  Username already exists",
+					  in.User.Name)
+	}
+
+	// Remove any duplicates
+	orgs := []*api_proto.Org{}
+	seen := map[string]bool{}
+	for _, org := range in.User.Orgs {
+		_, ok := seen[org.Id]
+		if !ok {
+			seen[org.Id] = true
+			orgs = append(orgs, org)
+		}
+	}
+	in.User.Orgs = orgs[:]
+
+	if !org_admin {
+		if !checkOrgMembership(in.User.Orgs, active_user.Orgs) {
+			return nil, status.Error(codes.PermissionDenied,
+				"User is not allowed to add users to orgs without the org admin role or being a member of all requested orgs.")
+		}
+		if !createUser {
+			removed := map[string]*api_proto.Org{}
+			for _, org := range user_record.Orgs {
+				removed[org.Id] = org
+
+			}
+			for _, org := range in.User.Orgs {
+				delete(removed, org.Id)
+			}
+			required_orgs := make([]*api_proto.Org, 0, len(removed))
+			idx := 0
+			for _, v := range removed {
+				required_orgs[idx] = v
+				idx += 1
+			}
+
+			if !checkOrgMembership(required_orgs, active_user.Orgs) {
+				return nil, status.Error(codes.PermissionDenied,
+					"User is not allowed to remove users from orgs without the org admin role or being a member of the org.")
+			}
+		}
+	}
+
+	org_manager, err := services.GetOrgManager()
+	if err != nil {
+		return nil, err
+	}
+
+	config_obj, err := org_manager.GetOrgConfig(services.ROOT_ORG_ID)
+	if err != nil {
+		return nil, err
+	}
+
+	err = setPassword(config_obj, in.User, in.Password, createUser)
+	if err != nil {
+		return nil, err
+	}
+
+	err = grantUserRoles(config_obj, &org_manager, in.User)
+	if err != nil {
+		return nil, err
+	}
+
+	user_record.Email = in.User.Email
+	user_record.Picture = in.User.Picture
+	user_record.VerifiedEmail = in.User.VerifiedEmail
+
+	// This must go after setting the password since SetPassword clears the Locked flag
+	user_record.Locked = in.User.Locked
+
+	err = users_manager.SetUser(ctx, user_record)
+	if err != nil {
+		return nil, err
+	}
+
+	msg := "users: Updating user record via API"
+	if createUser {
+		msg = "users: Creating user record via API"
+	}
+
+	logger := logging.GetLogger(org_config_obj, &logging.Audit)
+	logger.WithFields(logrus.Fields{
+		"Username":  active_user.Name,
+		"Principal": user_record.Name,
+	}).Info(msg)
+
+	return &emptypb.Empty{}, nil
+}
+
+func (self *ApiServer) CreateUser(ctx context.Context,
+				  in *api_proto.ChangeUserRequest) (*emptypb.Empty, error) {
+	return self.ChangeUser(ctx, in, true)
+}
+
+func (self *ApiServer) UpdateUser(ctx context.Context,
+				  in *api_proto.ChangeUserRequest) (*emptypb.Empty, error) {
+	return self.ChangeUser(ctx, in, false)
+}
+
+func (self *ApiServer) DeleteUser(
+	ctx context.Context,
+	in *api_proto.UserRequest) (*emptypb.Empty, error) {
+
+	users_manager := services.GetUserManager()
+	active_user, org_config_obj, err := users_manager.GetUserFromContext(ctx)
```

Comment:
More standard to call this "principal" is the user that is initiating the activity